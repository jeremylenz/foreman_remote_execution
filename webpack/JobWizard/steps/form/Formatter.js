import React, { useEffect } from 'react';
import { useSelector } from 'react-redux';
import { FormGroup, TextInput, TextArea } from '@patternfly/react-core';
import PropTypes from 'prop-types';
import SearchBar from 'foremanReact/components/SearchBar';
import { helpLabel } from './FormHelpers';
import { SelectField } from './SelectField';

const TemplateSearchField = ({
  name,
  controller,
  labelText,
  required,
  defaultValue,
  setValue,
  values,
}) => {
  const searchQuery = useSelector(
    state => state.autocomplete?.[name]?.searchQuery
  );
  useEffect(() => {
    setValue({ ...values, [name]: searchQuery });
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [searchQuery]);
  return (
    <FormGroup
      label={name}
      labelIcon={helpLabel(labelText, name)}
      fieldId={name}
      isRequired={required}
      className="foreman-search-field"
    >
      <SearchBar
        initialQuery={defaultValue}
        data={{
          controller,
          autocomplete: {
            id: name,
            url: `/${controller}/auto_complete_search`,
            useKeyShortcuts: true,
          },
        }}
        onSearch={() => null}
      />
    </FormGroup>
  );
};

export const formatter = (input, values, setValue) => {
  const isSelectType = !!input?.options;
  const inputType = input.value_type;
  const isTextType = inputType === 'plain' || !inputType; // null defaults to plain

  const { name, required, hidden_value: hidden } = input;
  const labelText = input.description;
  const value = values[name];

  if (isSelectType) {
    const options = input.options.split(/\r?\n/).map(option => option.trim());
    return (
      <SelectField
        key={name}
        isRequired={required}
        label={name}
        fieldId={name}
        options={options}
        labelText={labelText}
        value={value}
        setValue={newValue => setValue({ ...values, [name]: newValue })}
      />
    );
  }
  if (isTextType) {
    return (
      <FormGroup
        key={name}
        label={name}
        labelIcon={helpLabel(labelText, name)}
        fieldId={name}
        isRequired={required}
      >
        <TextArea
          className={hidden ? 'masked-input' : null}
          required={required}
          rows={2}
          id={name}
          value={value}
          onChange={newValue => setValue({ ...values, [name]: newValue })}
        />
      </FormGroup>
    );
  }
  if (inputType === 'date') {
    return (
      <FormGroup
        key={name}
        label={name}
        labelIcon={helpLabel(labelText, name)}
        fieldId={name}
        isRequired={required}
      >
        <TextInput
          placeholder="YYYY-mm-dd HH:MM"
          className={hidden ? 'masked-input' : null}
          required={required}
          id={name}
          type="text"
          value={value}
          onChange={newValue => setValue({ ...values, [name]: newValue })}
        />
      </FormGroup>
    );
  }
  if (inputType === 'search') {
    const controller = input.resource_type;
    // TODO: get text from redux autocomplete
    return (
      <TemplateSearchField
        key={name}
        name={name}
        defaultValue={value}
        controller={controller}
        labelText={labelText}
        required={required}
        setValue={setValue}
        values={values}
      />
    );
  }

  return null;
};

TemplateSearchField.propTypes = {
  name: PropTypes.string.isRequired,
  controller: PropTypes.string.isRequired,
  labelText: PropTypes.string,
  required: PropTypes.bool.isRequired,
  defaultValue: PropTypes.string,
  setValue: PropTypes.func.isRequired,
  values: PropTypes.object.isRequired,
};
TemplateSearchField.defaultProps = {
  labelText: null,
  defaultValue: '',
};
