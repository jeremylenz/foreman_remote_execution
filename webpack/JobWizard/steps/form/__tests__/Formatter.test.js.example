import React from 'react';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import * as patternfly from '@patternfly/react-core';
import { mount, shallow } from '@theforeman/test';
import { formatter } from '../Formatter';

jest.spyOn(patternfly, 'Select');
jest.spyOn(patternfly, 'SelectOption');
jest.spyOn(patternfly, 'FormGroup');
patternfly.Select.mockImplementation(props => <div props={props} />);
patternfly.SelectOption.mockImplementation(props => <div props={props} />);
patternfly.FormGroup.mockImplementation(props => <div props={props} />);
const mockStore = configureMockStore([]);
const store = mockStore({});

describe('formatter', () => {
  it('render date input', () => {
    const props = {
      name: 'date adv',
      required: false,
      options: '',
      advanced: true,
      value_type: 'date',
      resource_type: 'ansible_roles',
      default: '',
      hidden_value: false,
    };
    expect(shallow(formatter(props, {}, jest.fn()))).toMatchSnapshot();
  });
  it('render text input', () => {
    const props = {
      name: 'plain adv hidden',
      required: true,
      description: 'some Description',
      options: '',
      advanced: true,
      value_type: 'plain',
      resource_type: 'ansible_roles',
      default: 'Default val',
      hidden_value: true,
    };
    expect(shallow(formatter(props, {}, jest.fn()))).toMatchSnapshot();
  });
  it('render select input', () => {
    const props = {
      name: 'adv plain search',
      required: false,
      input_type: 'user',
      options: 'option 1\r\noption 2\r\noption 3\r\noption 4',
      advanced: true,
      value_type: 'plain',
      resource_type: 'ansible_roles',
      default: '',
      hidden_value: false,
    };
    expect(shallow(formatter(props, {}, jest.fn()))).toMatchSnapshot();
  });
  it('render search input', () => {
    const props = {
      name: 'search adv',
      required: false,
      options: '',
      advanced: true,
      value_type: 'search',
      resource_type: 'foreman_tasks/tasks',
      default: '',
      hidden_value: false,
    };
    expect(
      mount(
        <Provider store={store}>{formatter(props, {}, jest.fn())}</Provider>
      )
    ).toMatchSnapshot();
  });
});
