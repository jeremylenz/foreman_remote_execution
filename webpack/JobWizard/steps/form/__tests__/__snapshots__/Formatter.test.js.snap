// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formatter render date input 1`] = `
<div
  props={
    Object {
      "children": <ForwardRef
        className={null}
        id="date adv"
        onChange={[Function]}
        placeholder="YYYY-mm-dd HH:MM"
        required={false}
        type="text"
      />,
      "fieldId": "date adv",
      "isRequired": false,
      "label": "date adv",
      "labelIcon": null,
    }
  }
/>
`;

exports[`formatter render search input 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <TemplateSearchField
    controller="foreman_tasks/tasks"
    defaultValue=""
    key="search adv"
    labelText={null}
    name="search adv"
    required={false}
    setValue={
      [MockFunction] {
        "calls": Array [
          Array [
            Object {
              "search adv": undefined,
            },
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    }
    values={Object {}}
  >
    <mockConstructor
      className="foreman-search-field"
      fieldId="search adv"
      isRequired={false}
      label="search adv"
      labelIcon={null}
    >
      <div
        props={
          Object {
            "children": <SearchBar
              data={
                Object {
                  "autocomplete": Object {
                    "id": "search adv",
                    "url": "/foreman_tasks/tasks/auto_complete_search",
                    "useKeyShortcuts": true,
                  },
                  "controller": "foreman_tasks/tasks",
                }
              }
              initialQuery=""
              onSearch={[Function]}
            />,
            "className": "foreman-search-field",
            "fieldId": "search adv",
            "isRequired": false,
            "label": "search adv",
            "labelIcon": null,
          }
        }
      />
    </mockConstructor>
  </TemplateSearchField>
</Provider>
`;

exports[`formatter render select input 1`] = `
<mockConstructor
  fieldId="adv plain search"
  label="adv plain search"
>
  <mockConstructor
    className="without_select2"
    isOpen={false}
    isRequired={false}
    maxHeight="45vh"
    menuAppendTo={[Function]}
    onSelect={[Function]}
    onToggle={[Function]}
    selections={null}
  >
    <mockConstructor
      key="0"
      value="a"
    />
    <mockConstructor
      key="1"
      value="b"
    />
    <mockConstructor
      key="2"
      value="c"
    />
    <mockConstructor
      key="3"
      value="d"
    />
  </mockConstructor>
</mockConstructor>
`;

exports[`formatter render text input 1`] = `
<div
  props={
    Object {
      "children": <ForwardRef
        className="masked-input"
        id="plain adv hidden"
        onChange={[Function]}
        required={true}
        rows={2}
      />,
      "fieldId": "plain adv hidden",
      "isRequired": true,
      "label": "plain adv hidden",
      "labelIcon": <Popover
        aria-label="help-text"
        bodyContent="some Description"
        id="plain adv hidden-help"
      >
        <button
          aria-label="open-help-tooltip-button"
          className="pf-c-form__group-label-help"
          onClick={[Function]}
          type="button"
        >
          <HelpIcon
            color="currentColor"
            noVerticalAlign={true}
            size="sm"
          />
        </button>
      </Popover>,
    }
  }
/>
`;
